### PR TITLE
Add DatePartitionedLogger and RecordsWriter

### PR DESCRIPTION
**Pull request by frsv**
DatePartitionedLogger is a part of Avro logging hook taking care of managing the system paths and creating RecordWriter instances. The class mostly mirrors the [DatePartitionedLogger](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/tez/dag/history/logging/proto/DatePartitionedLogger.java) used in HiveProtoLoggingHook, excluding unnecessary methods and logic.

RecordsWriter is an Avro datum writer working with HDFS.

Opened the PR to `feature-records-constructor` for now. Depending on what is reviewed earlier I might to repoint it to main later.

### CODE REVIEW

**Review by dawidxc — APPROVED** [review]

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/DatePartitionedLogger.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  private static final FsPermission DIR_PERMISSION = FsPermission.createImmutable((short) 1023);
+  private final Path basePath;
+  private final Configuration conf;
+  private final Schema schema;
+  private final Clock clock;
+
+  public DatePartitionedLogger(Path baseDir, Configuration conf, Schema schema, Clock clock)
+      throws IOException {
+    this.conf = conf;
+    this.createDirIfNotExists(baseDir);
```
[the remark this answers is not in the document]
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/DatePartitionedLogger.java]: Done

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/DatePartitionedLogger.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    this.basePath = baseDir.getFileSystem(conf).resolvePath(baseDir);
+    this.schema = schema;
+    this.clock = clock;
+  }
+
+  public static LocalDate getDateFromDir(String dirName) {
+    if (!dirName.startsWith("date=")) {
```
[the remark this answers is not in the document]
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/DatePartitionedLogger.java]: Removed the prefix. I don't feel it's Hive-specific, at least hook wrote the data correctly to a folder without prefix.

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/DatePartitionedLogger.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      throw new IllegalArgumentException("Invalid directory: " + dirName);
+    } else {
+      return LocalDate.parse(dirName.substring(5), DateTimeFormatter.ISO_LOCAL_DATE);
+    }
+  }
+
+  public RecordsWriter getWriter(String fileName) throws IOException {
```
[the remark this answers is not in the document]
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/DatePartitionedLogger.java]: Done

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/DatePartitionedLogger.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  private void createDirIfNotExists(Path path) throws IOException {
+    FileSystem fileSystem = path.getFileSystem(this.conf);
+
+    try {
+      if (!fileSystem.exists(path)) {
+        fileSystem.mkdirs(path);
+        fileSystem.setPermission(path, DIR_PERMISSION);
+      }
+    } catch (IOException e) {
+      LOG.warn("Error while trying to set permission: ", e);
```
[the remark this answers is not in the document]
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/DatePartitionedLogger.java]: Done

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/RecordsWriter.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    dataFileWriter = new DataFileWriter<>(datumWriter);
+
+    FileSystem fs = filePath.getFileSystem(conf);
+
+    FSDataOutputStream outputStream = fs.create(filePath, true);
+    dataFileWriter.create(schema, outputStream);
+  }
+
+  public Path getPath() {
+    return this.filePath;
```
[the remark this answers is not in the document]
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/RecordsWriter.java]: Done

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/RecordsWriter.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  }
+
+  public void writeMessage(GenericRecord message) throws IOException {
+    dataFileWriter.append(message);
+  }
+
+  public void hflush() throws IOException {
```
[the remark this answers is not in the document]
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/RecordsWriter.java]: It's possible to call hflush() directly on the FSDataOutputStream. I have a backlog item to investigate whether the flush() on writer is just enough. Renamed for now

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/utils/SystemClock.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+ */
+package com.google.cloud.bigquery.dwhassessment.hooks.utils;
+
+import java.time.Clock;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+
+/** {@inheritDoc} Clock class implementation for getting the current system time in UTC. */
+public class SystemClock extends Clock {
```
[the remark this answers is not in the document]
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/utils/SystemClock.java]: Great, didn't know about it. Removed, thanks!

File: src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/DatePartitionedLoggerTest.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+  @Test
+  public void getDateFromDir_invalidDirectoryName_fail() {
+    IllegalArgumentException e =
+        assertThrows(
+            IllegalArgumentException.class, () -> DatePartitionedLogger.getDateFromDir("test"));
+
+    assertThat(e).hasMessageThat().isEqualTo("Invalid directory: test");
+  }
+
+  public static List<GenericRecord> readOutputRecords(HiveConf conf, String tmpFolder)
```
[the remark this answers is not in the document]
Comment (frsv) [reply src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/DatePartitionedLoggerTest.java]: It's actually a rebase leftover. Removed here, but changed in other places

File: src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/RecordsWriterTest.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    List<GenericRecord> records = readOutputRecords(conf, outputFilePath);
+    assertThat(records).containsExactly(message1, message2);
+  }
+
+  @Test
+  public void getPath_success() {
+    assertThat(recordsWriter.getPath()).isEqualTo(outputFilePath);
+  }
+
+  public static List<GenericRecord> readOutputRecords(HiveConf conf, Path outputFilePath)
```
[the remark this answers is not in the document]
Comment (frsv) [reply src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/RecordsWriterTest.java]: Done

File: src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/RecordsWriterTest.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+      throws IOException {
+    FileSystem fs = outputFilePath.getFileSystem(conf);
+    FSDataInputStream inputStream = fs.open(outputFilePath);
+
+    DatumReader<GenericRecord> reader = new GenericDatumReader<>(QUERY_EVENT_SCHEMA);
+
+    try (DataFileStream<GenericRecord> dataFileReader = new DataFileStream<>(inputStream, reader)) {
+      ArrayList<GenericRecord> records = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (frsv) [reply src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/RecordsWriterTest.java]: Done

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/DatePartitionedLogger.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+import org.apache.avro.Schema;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.permission.FsPermission;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Writes query events to files, partitioned by dates. */
+public class DatePartitionedLogger {
```
[the remark this answers is not in the document]
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/DatePartitionedLogger.java]: Renamed to `DatePartitionedRecordsWriterFactory`

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/RecordsWriter.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (dawidxc) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/RecordsWriter.java]: Thank you for renaming.

The method `hflush` on `FSDataOutputStream` does not take into account the data that might need to be flushed from `DatumWriter` first. At the same time it is not possible to call `hflush` on `DatumWriter`. Possibly calling both (i.e. first `DatumWriter#flush` and then `FSDataOutputStream#hflush`) might be a correct implementation of `hflush`.

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/DatePartitionedLogger.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (dawidxc) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/DatePartitionedLogger.java]: It is still called `logger` in `EventLogger`.

https://github.com/google/hadoop-migration-assessment-tools/blob/ed97cdbf06fa02c73bca11694202d98109b226c9/src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLogger.java#L44

[the remark this answers is not in the document]
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/DatePartitionedLogger.java]: Right, sorry. Fixed in #6 
Btw, new comments do not automatically unresolve conversation. It's hard to find the exact comment with all branches collapsed (something which Github could do better). If you can unresolve the conversation by yourself I would suggest doing that.

**Pull request merged**